dask_expr/_core.py

```diff
@@ -11,34 +10,59 @@
 import toolz
 from dask.dataframe.core import is_dataframe_like, is_index_like, is_series_like
 from dask.utils import funcname, import_required, is_arraylike
+from toolz.dicttoolz import merge
 
 from dask_expr._util import _BackendData, _tokenize_deterministic
 
 
-def _unpack_collections(o):
+def _unpack_expr(o):
     if isinstance(o, Expr):
-        return o
-
-    if hasattr(o, "expr"):
-        return o.expr
+        return o, o._name
+    elif hasattr(o, "expr"):
+        return o.expr, o.expr._name
     else:
-        return o
+        return o, None
 
 
 class Expr:
     _parameters = []
     _defaults = {}
 
     def __init__(self, *args, **kwargs):
+        self._dependencies = {}
         operands = list(args)
         for parameter in type(self)._parameters[len(operands) :]:
             try:
                 operands.append(kwargs.pop(parameter))
             except KeyError:
                 operands.append(type(self)._defaults[parameter])
         assert not kwargs, kwargs
-        operands = [_unpack_collections(o) for o in operands]
-        self.operands = operands
+        parsed_operands = []
+        children = set()
+        _subgraphs = []
+        _subgraph_instances = []
+        _graph_instances = {}
+        for o in operands:
+            expr, name = _unpack_expr(o)
+            parsed_operands.append(expr)
+            if name is not None:
+                children.add(name)
+                _subgraphs.append(expr._graph)
+                _subgraph_instances.append(expr._graph_instances)
+                _graph_instances[name] = expr
+
+        self.operands = parsed_operands
+        name = self._name
+        # Graph instances is a mapping name -> Expr instance
+        # Graph itself is a mapping of dependencies mapping names to a set of names
+        self._graph_instances = merge(_graph_instances, *_subgraph_instances)
+        self._graph = merge(*_subgraphs)
+        self._graph[name] = children
+        # Probably a bad idea to have a self ref
+        self._graph_instances[name] = self
+
+    def __hash__(self):
+        raise TypeError("Don't!")
```

Comment:
"Funny" story: The above implementation is maintaining two dicts. One that maps names and dependencies and one that just maps names to the actual objects. Initially I just tried to use the Expr objects themselves in the mappings using a mapping from `Expr -> set[Expr}`. That triggered funny recursion errors and it took me a while to understand that...

Whenever there is a hash collision, the implementation of set falls back to use `__eq__` to compare the new object with the one that is stored under a given hash, i.e. `object.__eq__(old, new)`. However, this doesn't evaluate to a bool but defines another expression... 💥

Reply:
Yeah, definite `__eq__` is pretty hazardous in Python.  It would be reasonable, I think, to drop `Expr.__eq__` and use explicit `Eq(a, b)` calls instead, and then rely on `Frame.__eq__` for user comfort.

Reply:
Or rather, use whatever class is defined in `collections.py` to hold the `__eq__` method